instrumentation/opentelemetry_finch/lib/opentelemetry_finch.ex

```diff
@@ -0,0 +1,73 @@
+defmodule OpentelemetryFinch do
+  @moduledoc """
+  OpentelemetryFinch uses [telemetry](https://hexdocs.pm/telemetry/) handlers to
+  create `OpenTelemetry` spans.
+
+  ## Usage
+
+  In your application start:
+
+      def start(_type, _args) do
+        OpentelemetryFinch.setup()
+
+        # ...
+      end
+
+  """
+
+  require OpenTelemetry.Tracer
+
+  @typedoc "Setup options"
+  @type opts :: []
+
+  @doc """
+  Initializes and configures the telemetry handlers.
+  """
+  @spec setup(opts()) :: :ok
+  def setup(_opts \\ []) do
+    :telemetry.attach(
+      {__MODULE__, :request_stop},
+      [:finch, :request, :stop],
+      &__MODULE__.handle_request_stop/4,
+      %{}
+    )
+  end
+
+  @doc false
+  def handle_request_stop(_event, measurements, meta, _config) do
+    duration = measurements.duration
+    end_time = :opentelemetry.timestamp()
+    start_time = end_time - duration
+
+    status =
+      case meta.result do
+        {:ok, response} -> response.status
+        _ -> 0
+      end
+
+    attributes = %{
+      "http.scheme": meta.request.scheme,
+      "http.host": meta.request.host,
+      "http.port": meta.request.port,
+      "http.path": meta.request.path,
+      "http.method": meta.request.method,
+      "http.status": status,
+    }
+
+    s =
+      OpenTelemetry.Tracer.start_span("client http", %{
```

Comment:
If you start the span at the end then it won't be able to be propagated as trace context in the headers.

Reply:
Thank you very much, I have a question, something like this?

```elixir
     parent_context = OpentelemetryProcessPropagator.fetch_parent_ctx(1, :"$callers")

     if parent_context != :undefined do
       OpenTelemetry.Ctx.attach(parent_context)
     end

     s =
       OpenTelemetry.Tracer.start_span(client http, %{ ...
```

Reply:
Na, I mean calling `start_span` in Finch's request `start` event instead of in `stop`.

Reply:
I understand. Since this instrumentation uses telemetry handlers to create OpenTelemetry spans from Finch events, when you want to propagate the trace context in the headers, before making the request, the headers should be set as follows:

```elixir
headers = :otel_propagator_text_map.inject([])

Finch.build(:get, url, headers)
  |> Finch.request(HttpFinch)
```

Do you think that if I put the above information in the documentation, the instrumentation can be a candidate for merge?

Thanks for your time

Reply:
Ah, blergh, I don't know what I was thinking.

What should happen is the injected trace context is the finch generated span, not its parent. But I guess that isn't possible because telemetry events are being used.

We might just have to settle for this. @bryannaegele ?

The attributes to match the semantic conventions just need to be updated.

Reply:
Also, yes, including the `inject` in the readme would be good. But it should also note in there that this won't give the trace tree that is expected because the children in the service the request is sent to will be of the parent of the finch request instead of the finch request itself.

This will need to be improved somehow but I guess doesn't need to block this PR. I guess there will need to be a Finch wrapper module, unless Finch incorporated a form of middlewares.

Reply:
I just updated the readme. Thank you very much